hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/metrics/MetricsContext.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.core.metrics;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Context object which is able to provide localized counter instances to
+ * be used as sources of custom metrics. Each {@link com.hazelcast.jet.core.Processor}
+ * will have its own instance. Hence user metrics have the same granularity
+ * as Processors.
+ */
+public interface MetricsContext {
+
+    /**
+     * Return a counter that can be used to set values for a custom, per
+     * Processor metric.
+     */
+    @Nonnull
+    AtomicLong getCounter(String name);
```

Comment:
why AtomicLong? isn't it too low level?

Reply:
I think it has to be something thread safe... Should I put some kind of generic wrapper on top of it?

Reply:
yes, it should be wrapped. we can also support a lambda.

Reply:
What you said here makes sense to me, but let's say I add a method like `setCounterSource(String name, SupplierEx<Long> supplier)` to `MetricsContext` which would facilitate the user providing a lambda, how would I enforce thread safety? We do such things internally, but then we control the threads. Here we don't though... Have to think about it a bit more.

Reply:
Have pushed **A** solution, which hides the `AtomicLong` and allows both for explicit setting of values and implicit provision of lambda sources, but it's not perfect (in particular I mean that I wrote a comment that the lambda has to be thread-safe, but we can't enforce it). Take a look when you can and we'll see if it's acceptable.